### PR TITLE
322 - remove CURRENT_TENANT_THEN_PROVIDER

### PR DESCRIPTION
**Pull request by rnewbigging**
Removed deprecated `ScpCfDestinationRetrievalStrategy.CURRENT_TENANT_THEN_PROVIDER`.

## Definition of Done

- [x] Functionality scope stated & covered
- [x] Tests cover the scope above
- [x] Documentation updated
- [x] Release notes updated

### CODE REVIEW

File: cloudplatform/cloudplatform-connectivity-scp-cf/src/test/java/com/sap/cloud/sdk/cloudplatform/connectivity/ScpCfDestinationLoaderTest.java [hunk 1]
```diff
@@ -565,82 +564,6 @@ public void testLoadAllDestinationsProviderOnly()
             .containsOnly("CC8-HTTP-BASIC", "CC8-HTTP-CERT", "CC8-HTTP-CERT1");
     }
 
-    @SuppressWarnings( "deprecation" )
-    @Test
-    public void testLoadAllDestinationsWithProviderFallback()
```
Comment (newtork) [inline cloudplatform/cloudplatform-connectivity-scp-cf/src/test/java/com/sap/cloud/sdk/cloudplatform/connectivity/ScpCfDestinationLoaderTest.java:570]: (Minor)

I would be careful removing this test (case). From method name/description it seems to be a reasonable scenario(?). I'll check whether we can restructure the test code, while maintaining the test intent.

File: cloudplatform/cloudplatform-connectivity-scp-cf/src/test/java/com/sap/cloud/sdk/cloudplatform/connectivity/ScpCfDestinationLoaderTest.java [hunk 2]
```diff
@@ -749,67 +672,6 @@ public void testDestinationRetrievalSubscriberOnly()
         assertThat(loadedHttpDestination.getUri()).isEqualTo(URI.create(subscriberUrl));
     }
 
-    @SuppressWarnings( "deprecation" )
-    @Test
-    public void testDestinationRetrievalWithProviderFallback()
```
Comment (newtork) [inline cloudplatform/cloudplatform-connectivity-scp-cf/src/test/java/com/sap/cloud/sdk/cloudplatform/connectivity/ScpCfDestinationLoaderTest.java:754]: (Minor)

I would be careful removing this test (case). From method name/description it seems to be a reasonable scenario(?). I'll check whether we can restructure the test code, while maintaining the test intent.

File: cloudplatform/cloudplatform-connectivity-scp-cf/src/main/java/com/sap/cloud/sdk/cloudplatform/connectivity/DestinationRetrievalStrategyResolver.java [hunk 3]
```diff
@@ -164,34 +163,25 @@ private ScpCfDestinationTokenExchangeStrategy getDefaultTokenExchangeStrategy()
...
-        if( tokenExchangeStrategy != LOOKUP_THEN_EXCHANGE && retrievalStrategy != CURRENT_TENANT_THEN_PROVIDER ) {
+        if( tokenExchangeStrategy != LOOKUP_THEN_EXCHANGE ) {
             strategy = resolveSingleRequestStrategy(retrievalStrategy, tokenExchangeStrategy);
             return () -> destinationRetriever.apply(strategy);
-        }
-
-        // deal with LOOKUP_THEN_EXCHANGE but exclude CURRENT_TENANT_THEN_PROVIDER for now
-        if( retrievalStrategy != CURRENT_TENANT_THEN_PROVIDER ) {
+        } else {
+            // LOOKUP_THEN_EXCHANGE
```
Comment (newtork) [inline cloudplatform/cloudplatform-connectivity-scp-cf/src/main/java/com/sap/cloud/sdk/cloudplatform/connectivity/DestinationRetrievalStrategyResolver.java:184]: (Minor)

Code style could be improved, i.e. handle lookup-then-exchange in the if-branch above with early-return and use the simple-case without else block to improve readability. But very minor.

File: cloudplatform/cloudplatform-connectivity-scp-cf/src/main/java/com/sap/cloud/sdk/cloudplatform/connectivity/DestinationRetrievalStrategyResolver.java [hunk 4]
```diff
@@ -205,9 +195,6 @@ Supplier<ScpCfDestinationServiceV1Response> prepareSupplier(
                 return destinationRetriever.apply(tokenExchangeOnlyStrategy);
             };
         }
-
-        // handle CURRENT_TENANT_THEN_PROVIDER where current tenant != provider
-        return prepareSupplierForSubscriberThenProviderCase(tokenExchangeStrategy);
     }
 
     Supplier<ScpCfDestinationServiceV1Response> prepareSupplierForSubscriberThenProviderCase(
```
Comment (newtork) [inline cloudplatform/cloudplatform-connectivity-scp-cf/src/main/java/com/sap/cloud/sdk/cloudplatform/connectivity/DestinationRetrievalStrategyResolver.java:200]: (Major/Question)

Can this method be removed?

File: cloudplatform/cloudplatform-connectivity-scp-cf/src/main/java/com/sap/cloud/sdk/cloudplatform/connectivity/DestinationRetrievalStrategyResolver.java [hunk 5]
```diff
@@ -353,30 +302,12 @@ Supplier<List<Destination>> prepareSupplierAllDestinations( @Nonnull final Desti
...
-                                "Falling back to the provider tenant after failing to retrieve destinations for the subscriber tenant.");
-                        log.debug("Lookup of all destinations for the subscriber tenant failed.", e);
-                    }
-                    return allDstinationRetriever.apply(TECHNICAL_USER_PROVIDER);
-                };
+                return () -> allDestinationRetriever.apply(TECHNICAL_USER_PROVIDER);
             }
             case ONLY_SUBSCRIBER:
             case CURRENT_TENANT:
             default: {
```
Comment (newtork) [inline cloudplatform/cloudplatform-connectivity-scp-cf/src/main/java/com/sap/cloud/sdk/cloudplatform/connectivity/DestinationRetrievalStrategyResolver.java:309]: (Minor/Unrelated)

Default case could be throwing an `IllegalStateException` or `IllegalArgumentException`.

File: release_notes_next_major.md [hunk 6]
```diff
@@ -102,6 +102,7 @@ blog: https://blogs.sap.com/?p=xxx
 - Removed the following elements from enum `com.sap.cloud.sdk.cloudplatform.connectivity.AuthenticationType`:
   - `APP_TO_APP_SSO` 
   - `INTERNAL_SYSTEM_AUTHENTICATION` 
+- Removed the deprecated enum element `CURRENT_TENANT_THEN_PROVIDER` from `com.sap.cloud.sdk.cloudplatform.connectivity.ScpCfDestinationRetrievalStrategy`.
```
Comment (newtork) [inline release_notes_next_major.md:105]: (Minor)
We could add another hint to the release notes 🤔 

```suggestion
- Removed the deprecated enum element `CURRENT_TENANT_THEN_PROVIDER` from `com.sap.cloud.sdk.cloudplatform.connectivity.ScpCfDestinationRetrievalStrategy`.
  As alternative please run the destination lookup queries separately in your application for `ONLY_SUBSCRIBER` and `ALWAYS_PROVIDER` as fallback.
```

File: cloudplatform/cloudplatform-connectivity-scp-cf/src/test/java/com/sap/cloud/sdk/cloudplatform/connectivity/ScpCfDestinationLoaderTest.java (hunk 1 above)
Comment (newtork) [reply cloudplatform/cloudplatform-connectivity-scp-cf/src/test/java/com/sap/cloud/sdk/cloudplatform/connectivity/ScpCfDestinationLoaderTest.java:570]: Sorry, there is no fallback logic if current-tenant-then-provider is gone.

File: cloudplatform/cloudplatform-connectivity-scp-cf/src/test/java/com/sap/cloud/sdk/cloudplatform/connectivity/ScpCfDestinationLoaderTest.java (hunk 2 above)
Comment (newtork) [reply cloudplatform/cloudplatform-connectivity-scp-cf/src/test/java/com/sap/cloud/sdk/cloudplatform/connectivity/ScpCfDestinationLoaderTest.java:754]: Sorry, there is no fallback logic if current-tenant-then-provider is gone.

File: cloudplatform/cloudplatform-connectivity-scp-cf/src/main/java/com/sap/cloud/sdk/cloudplatform/connectivity/DestinationRetrievalStrategyResolver.java (hunk 4 above)
Comment (rnewbigging) [reply cloudplatform/cloudplatform-connectivity-scp-cf/src/main/java/com/sap/cloud/sdk/cloudplatform/connectivity/DestinationRetrievalStrategyResolver.java:200]: Ah yes, I meant to come back to this after cleaning up the tests. Thanks for the catch.

File: cloudplatform/cloudplatform-connectivity-scp-cf/src/main/java/com/sap/cloud/sdk/cloudplatform/connectivity/DestinationRetrievalStrategyResolver.java [hunk 7]
```diff
@@ -298,17 +241,21 @@ Supplier<List<Destination>> prepareSupplierAllDestinations( @Nonnull final Desti
...
             }
             case ONLY_SUBSCRIBER:
-            case CURRENT_TENANT:
-            default: {
+            case CURRENT_TENANT: {
                 return () -> allDestinationRetriever.apply(TECHNICAL_USER_CURRENT_TENANT);
             }
+            default: {
+                throw new IllegalArgumentException(
+                    "The provided destination retrieval strategy " + strategy + "is not valid.");
```
Comment (newtork) [inline cloudplatform/cloudplatform-connectivity-scp-cf/src/main/java/com/sap/cloud/sdk/cloudplatform/connectivity/DestinationRetrievalStrategyResolver.java:257]: Minor typo
```suggestion
                    "The provided destination retrieval strategy " + strategy + " is not valid.");
```

**Review by newtork — APPROVED** [review]

**Pull request merged**